File: C++/student.cpp
#include <iostream>
#include <string>

using namespace std;

class Teacher{
private:
	string name;
public:
	Teacher(string teachName){
		name = teachName;
	}
	Teacher( ){
		name = " ";
	}
	void setTeachName(string teachName){
		name = teachName;
	}
	string getTeachName(void){
		return name;
	}
	void readTeacher( ){
		cout << "Enter teacher name: ";
		getline(cin,name);
	}
	void printTeacher(){
		cout << "Name: " << name;
	}
};

class Subject{
  private:
  		string name,code;
  		Teacher teacher;
  public:
     Subject( string subjName,string subjCode,Teacher teacher ){
     	name = subjName;
     	code = subjCode;
     	this->teacher = teacher;
     }	
     Subject(){
     	name = "Student Name";
     	code = "Subject Code";
     	this->teacher = teacher;
     }	
     string getSubjName(void){
     	return name;
     }
     string getSubjCode(void){
     	return code;
     }
     void readSubject(){
     	cout << "Enter Subject: ";
     	getline(cin,name);
     	cout << "Enter Code: ";
     	getline(cin,code);
     }
     void printSubject(Teacher teacher){
     	cout << name <<"\t"<< code <<"\t"<< teacher.getTeachName() <<endl;
     }
};

 class Marks{
 private:
 	float mark1,mark2,avgMark;
 	Subject subj;
 	Teacher teacher;
 public:
 	Marks(float mk1,float mk2){
 		mark1 = mk1;
 		mark2 = mk2;
 		avgMark = (mk1 + mk2)/2;
 	}
 	Marks(){
 		mark1 = 0.0;
 		mark2 = 0.0;
 		if( mark1==0 || mark2==0){
 			avgMark = 0;
 		}else{
 			 avgMark = (mark1 + mark2)/2;
         }
    }     
 	float getMark1(void){
 		return mark1;
 	}
 	float getMark2(void){
 		return mark2;
 	}
 	float getAverageMark(void){
 		return avgMark;
 	}
 	void readMarks(Subject subj){
 	     cout << "Subject: " << subj.getSubjName() <<endl;
         cout << "Enter first Mark: ";
         cin >> mark1;
         cout << "Enter second Mark: ";
         cin >> mark2;
 	}
 };

class Student{
private:
	string firstName,lastName,gender,dateOfBirth;
	int admisNumber;
protected:
	Teacher teacher;
	Marks mark;
	Subject subj;
public:
	Student(string fname,string lname,string gen,string DoB,int number){
         firstName = fname;
         lastName = lname;
         gender = gen;
         dateOfBirth = DoB;
         admisNumber = number;
	}
	Student(){
		firstName = "First Name";
		lastName = "Last Name";
		gender = "Gender";
		dateOfBirth = "01-January-19";
		admisNumber = 0;
	}
	string getName(void){
		return firstName+" "+lastName;
	}
	string getGender(void){
		return gender;
	}
	string getDatofBirth(void){
		return dateOfBirth;
	}
	int getAdminNumber(void){
		return admisNumber;
	}
	void studentInfo(){
		//cin.ignore();
		cout << "Enter First Name: ";
		getline(cin,firstName);
		cout << "Enter Last Name: ";
		getline(cin,lastName);
		cout << "Enter Gender: ";
		getline(cin,gender);
		cout << "Enter Date Of Birth: ";
		getline(cin,dateOfBirth);
		cout << "Enter Admission Number: ";
		cin >> admisNumber;
	}
	void showStudent(){
		cout << "Name: " << getName() <<"\t"<< "Gender: " << gender <<endl;
		cout << "Date of Birth: " << dateOfBirth <<"\t"<< "Admission Number: "<< admisNumber <<endl;
	}
	void printStudent(){
         cout << getName() <<"\t"<< gender <<"\t"<< dateOfBirth <<"\t"<< admisNumber << endl;
	}
	void readReport( Marks mark,Subject subj){	
	  cout << getName() << " " << getGender() <<endl;
	  mark.readMarks(subj);
	}
	void printReport(Subject subj,Teacher teacher,Marks mark){
		showStudent();
		cout << subj.getSubjName() <<"\t"<< teacher.getTeachName() <<"\t"<< mark.getMark1() << "\t";
		cout << mark.getMark2() <<"\t"<<mark.getAverageMark() <<endl; 
	}
};

int main(){
	Teacher teach("Teacher");
	Subject subj("Subject","SUB203",teach);
	Marks mark(0,0);
	Student stud("Student","Stude","Male","24-March-19",32552);
	cout << "*************TEACHER INFORMATION********************"<<endl;
	teach.readTeacher();
	cout <<"\n\n";
	cout << "******************SUBJECT INFORMATION*****************" <<endl;
	subj.readSubject();
	cout <<"\n\n";
	cout << "**************STUDENT INFORMATION*********************" <<endl;
	stud.studentInfo();
	cout <<"\n\n";
	stud.readReport(mark,subj);
	cout << "***********SUBJECT AND TEACHERS**********************" <<endl;
	subj.printSubject(teach);
	cout << "\n\n";
	cout << "**********STUDENT AND MARKS**************************" << endl;
	stud.printReport(subj,teach,mark);
	cout << "\n\n";
	return 0;
}

File: PYTHON/student.py
class Teacher:
	def __init__(self, name,dateEmployed):
		self.name = name
		self.dateEmployed = dateEmployed
	def Read(self):
	   name = input("Enter Teacher Name: ")
	   dateEmployed = input("Date Employed: ")	 
	def PrintTeacher(self):
	   print("Name: ",self.name," Date Employed: ",self.dateEmployed)	

class Subject:
	teacher = Teacher("Chefor","24-march-26")
	def __init__(self,name,code,teacher):
		self.name = name;
		self.code = code;
		self.teacher = teacher;
	def Read(self,teacher):
	    name = input("Enter Subject Name: ")
	    code = input("Enter Subject Code: ")
	    print("Teacher is ",teacher.name);
	def PrintSubject(self):
	     print("Subject: ",self.name,"Code: ",self.code,"Teacher: ",teacher.name)  	      

class Student:
	def __init__(self,firstName,lastName,gender,dateOfBirth,matricule):
		    self.firstName = firstName
		    self.lastName = lastName
		    self.gender = gender
		    self.dateOfBirth = dateOfBirth
		    self.matricule = matricule

	def Read(self):
	   self.firstName = input("Enter Student First Name: ")
	   self.lastName = input("Enter Student Last Name: ")
	   self.gender = input("Enter Gender: ")
	   self.dateOfBirth = input("Enter Student Date of Date: ")
	   self.matricule = int(input("Enter Student Matricule: "))

	def PrintStudent(self):
	   print("Name: ",self.firstName," ",self.lastName)
	   print("Gender:",self.gender)
	   print("Date Of Birth: ",self.dateOfBirth)
	   print("Matricule: ",self.matricule) 
#Student mark classs	   
class Marks:
    stud = Student("FirstName","LastName","Gender","01-January-19",0)
    subj = Subject("Subject","Code",teacher) 

    def __init__(self,mark1,mark2,avgMark):
           self.mark1 = mark1
           self.mark2 = mark2
           self.avgMark = (self.mark1 + self.mark2) / 2

    def ReadMark(self,subj,stud):
  	   print("****************************SUBJECT**************************")
  	   print("Subject: ",subj.name,"Code: ",subj.code,"Teacher: ",subj.teacher.name)	
  	   print("==============================================================")
  	   print("*************************STUDENT******************************")
  	   print(stud.firstName," ",stud.lastName," ",stud.gender)
  	   print("--------------------------------------------------------------")
  	   self.mark1 = float(input("Enter Firs Sequence Mark: "))
  	   self.mark2 = float(input("Enter Second Sequence Mark: "))
  	   self.avgMark = (self.mark1 + self.mark2)/2;

    def PrintStudentReport(self):
       #print("\n\n***********************STUDENT REPORT***********************")	
  	   print(stud.firstName," ",stud.lastName,"Date of Birth: ",stud.dateOfBirth)
  	   print(stud.gender)
  	   print("Matricule: ",stud.matricule)
  	   print("---------------------------------------------------------------")
  	   print("===============================================================")
  	   print(subj.name," ",subj.teacher.name," ",subj.code,"  ",self.mark1,"  ",self.mark2," ",self.avgMark)


#Creating Teacher Object
teacher = Teacher("Chefor","24-03-29")
#Reading teachers information
teacher.Read()
#Printing Teachers Information
print("*********************TEACHER*********************")
teacher.PrintTeacher()

#Creating subject Object
subj = Subject("Subject","Code",teacher)
#Reading the subject information
subj.Read()
#Printing the subjects
subj.PrintSubject()	 
#Creating Student Object
stud = Student("FirstName","LastName","Gender","01-January-19",1)
#Reading students Information
stud.Read()
#print Students Information
stud.PrintStudent()

#Marks object
mark = Marks(1,1,0)  
#Reading Students marks
mark.ReadMark()
#Printing students Reports
print("\n\n***********************STUDENT REPORT*********************")
mark.PrintStudentReport()

		